pyomo/contrib/doe/fim_doe.py

```diff
@@ -6,13 +6,131 @@
+        '''
+        # check if measurement variables need to be flattened
+        self.name_and_index = self.name_and_index_generator(self.measurement_all_info)
+        self.flatten_measure_name = self.generate_flatten_name(self.name_and_index)
+        self.flatten_measure_timeset = self.generate_flatten_timeset(self.measurement_all_info,self.flatten_measure_name, self.name_and_index)
+        print('All measurements are flattened.')
+        print('Flatten measurement name:', self.flatten_measure_name)
+        print('Flatten measurement timeset:', self.flatten_measure_timeset)
+        return self.flatten_measure_name, self.flatten_measure_timeset
+
+    def generate_flatten_name(self, measure_name_and_index):
+        '''Generate measurement flattened names
+        Parameters
+        ----------
+        measure_name_and_index: a dictionary, keys are measurement names, values are lists of extra indexes
+
+        Returns
+        ------
+        jac_involved_name: a list of flattened measurement names
+        '''
+        flatten_names = []
+        for j in list(measure_name_and_index.keys()):
+            if measure_name_and_index[j] is not None: # if it has extra index
+                for ind in measure_name_and_index[j]:
+                    flatten_name = j + '_index_' + str(ind)
+                    flatten_names.append(flatten_name)
+            else:
+                flatten_names.append(j)
+
+        return flatten_names
+
+    def generate_flatten_timeset(self, all_info, flatten_measure_name,name_and_index):
+        '''
+        Generate flatten variables timeset. Return a dict where keys are the flattened variable names,
+        values are a list of measurement time.
+        '''
+        flatten_measure_timeset = {}
+        for i in flatten_measure_name:
+            # split the flattened name if needed
+            if '_index_' in i:
+                measure_name = i.split('_index_')[0]
+                measure_index = i.split('_index_')[1]
+                if type(name_and_index[measure_name][0]) is int:
+                    measure_index = int(measure_index)
+                flatten_measure_timeset[i] = all_info[measure_name][measure_index]
+            else:
+                flatten_measure_timeset[i] = all_info[i]
+        return flatten_measure_timeset
+
+    def check_subset(self,subset):
+        '''
+        Check if the subset is correctly defined with right name, index and time.
+        TODO: return True or False . Optional argument: throw error = False
+        subset: measurement name and index involved in jacobian calculation
+        '''
+        flatten_subset = self.generate_flatten_name(subset)
+        flatten_name_index = self.name_and_index_generator(subset)
+        flatten_timeset = self.generate_flatten_timeset(subset, flatten_subset, flatten_name_index)
+        for i in flatten_subset:
+            if i not in self.flatten_measure_name:
```

Comment:
Add comments. Verify every flattened measurement name in `subset` is contained in `self`.